### PR TITLE
Ads: handle setting ad position

### LINKED ISSUE #237: Ads: configurable placement
#233 Introduces ads, but they are appended to the bottom of the newsletter email only. Next step is to have options to place an ad at a specific position in the email newsletter.

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #237. Also, fixes a small visual bug in the ad removal modal.

### How to test the changes in this Pull Request:

1. Add some newsletter ads, set positions in sidebar settings
2. Observe that an ad has 100% position (bottom of content) by default
3. Observe that the ads are placed accordingly in newsletter HTML
4. _Additional thing_: Click "Delete" on an ad in ad manager to remove it. Observe the buttons in the modal have some spacing between.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: src/ads-admin/editor/index.js [hunk 1]
```diff
@@ -30,8 +30,15 @@ const AdEdit = ( { expiryDate, editPost } ) => {
...
-				title={ __( 'Expiry date', 'newspack-newsletters' ) }
+				title={ __( 'Ad settings', 'newspack-newsletters' ) }
 			>
+				<RangeControl
+					label={ __( 'Approximate position (in percent)' ) }
+					value={ positionInContent }
+					onChange={ position_in_content => editPost( { meta: { position_in_content } } ) }
+					min={ 0 }
+					max={ 100 }
+				/>
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply src/ads-admin/editor/index.js:51]: Added in d01de4cdbe0c8f1a535d1693857a44113a078af3

File: includes/class-newspack-newsletters-renderer.php [hunk 2]
```diff
@@ -556,38 +556,62 @@ private static function post_to_mjml_components( $post, $include_ads ) {
...
-			$body .= $ads_markup;
+		}
+
+		// Build MJML body and insert ads.
+		$current_position = 0;
+		foreach ( $valid_blocks as $block ) {
+			$block_content     = self::render_mjml_component( $block );
+			$current_position += strlen( wp_strip_all_tags( $block_content ) );
+			foreach ( $ads_to_insert as &$ad_to_insert ) {
+				if ( ! $ad_to_insert['is_inserted'] && $current_position > $ad_to_insert['precise_position'] ) {
```
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/class-newspack-newsletters-renderer.php]: The issue here is that the layout used in this testing ("Support") has just three top-level blocks – all the paragraphs are in a group block. The insertion logic goes block by block, in order to insert the ads between top-level blocks. This way an ad will not be place mid-block, causing issues. 
That said, the insertion logic needed a tweak anyway: a4e11e6b6b7752e50257f00edbeeb82df4b3bed7

**Comment by adekbadek** [comment]
> I don't love how the ads are created and edited completely separately from the email content.

> we should try to render an uneditable ad placeholder in the editor approximately where an ad will appear in the content.

> In my ideal scenario, active ads would automatically appear as blocks in their default positions for each new newsletter

These are interesting suggestions and I agree implementing them would improve the UX. Added #266 – feel free to edit it.

> I also don't think active ads should necessarily always appear in every newsletter. 

Great idea, added in e0b2c7e43f701af70e198ea2af79e87131165329

**Comment by dkoo** [comment]
After discussing with @adekbadek, I think we need to tweak the insertion logic so that ads can appear in between blocks inside of a Group block. Right now all four default layouts use Group blocks to wrap most content, making the ad positioning look off compared to the percentage setting for each ad. I think we need to go one level deeper than Group so that the positioning appears closer to the percentage within all content.

**Comment by adekbadek** [comment]
@dkoo I've added in-group-block ad insertion in 2393578, please check it out.

File: src/newsletter-editor/sidebar/index.js [hunk 3]
```diff
@@ -93,16 +94,34 @@ const Sidebar = ( {
 		</Fragment>
 	);
 
+	const updateMetaValue = ( key, value ) => {
+		editPost( { meta: { [ key ]: value } } );
```
Comment (dkoo) [inline src/newsletter-editor/sidebar/index.js:98]: @adekbadek Love this pattern, I might steal it for future work 😄

**Pull request merged**